doc/rntrack.1
.TH RNTRACK 1 "13 January 2005"
.SH NAME
RNtrack \- mail tracker and packer for FTN-networks
.SH SYNOPSIS
.B rntrack
[ options ]
.SH DESCRIPTION
.I  RNtrack
is so called netmail "tracker". It's puprose is to process netmail on
netmail hubs, though it can be useful for an ordinary node. It takes a letter,
compares it with a mask and, if comparison was successful, it executes
corresponding operation with this letter.
.SH OPTIONS
.TP
.B \-c configfile
Use alternative config file. Space between -c and the beginning of path is obligatory. Quotation
marks are necessary. This parameter overrides 
.B RNTRACKCONFIG
environment variable.
.TP
.B \-u
Unpack outbound. Shows that outbound repack is necessary
Before execution of all main functions, described in configuration
file, RNtrack will repack all outbound to a directory, defined by
TempMail: keyword.
.TP
.B \-t
Do ScanDir: with expired time of execution (instructions for which single execution is
shown, without respect of whether or not they should be executed right now).
.TP
.B \-h, \-?
Display help screen.
.SH ENVIRONMENT
.TP
.B RNTRACKCONFIG
Use alternative config file.
.SH SUPPORTED PLATFORMS
Linux, FreeBSD, OpenBSD, NetBSD, Solaris, MacOS X, AIX, OS/2, DOS and Win32.
.SH AUTHOR
Alex Soukhotine and others

ChangeLog

! - Bug fixed
# - Added feature or changes

1.12 22-10-05
-------------

! Fix build with mingw32
! Fix Solaris/x86 byte order (thanks to Alex V. Bagmanov)
# Added Immediate outbound flavour
# New keyword "Domain"
# Added The Brake! fileboxes and outbound support
# Added personal fileboxes support
# String processing code was replaced (taken from OpenBSD source).
! CVS version string format replaced to: RNtrack X.Y.YYYYMMDD/OS.
# Added 'rntrack-stable' cvs branch. This source tree will contain only bugfixes
  and tested features from current source.
# New action: AddKludge
# Add new mask flag "+M/-M" (check message size). Removed automatical
  message size check when MaxMsgSize set.
# Rewritten Watcom makefiles. Now you don't need build smapi manually. Now you can use
  ENABLE_LOG_PID, ENABLE_SYSLOG_LOG_FORMAT and CONFIG make options.
! Temporary disabled @MsgLoops, @MaxAge and @MaxAttachSize macroses.
# Test build for Linux/x86_64

1.11 30-01-05 (aka as11)
------------------------

# RNtrack: fork from FTrack 3.1.1-as 2005/01/12 03:00.
! Fix FreeBSD port (thanks to Andrey Slusar, 2:467/126)
! Fix compilation with gcc 3.4
# Added Slackware linux package build script (MakeFiles/linux/slackware)
# New variable: MaxPktSize.
! gcc 3.4 and FreeBSD port fixes (thanks to Andrey Slusar, 2:467/126)
# New perl API function: NewMsg() which creates message and put it to base.
# For security reasons default value of BadPackets changed to 'Skip'.
! Upgrade to cvs smapi-2_4-stable.
! Fix JAM-base (thanks to Michael Kostylev, 2:5030/1530)
! Fix OADDRESS subfield in JAM-base (thanks to Michael Kostylev, 2:5030/1530)
# Now we have own FTSC product code: 16FF
# New variable: MaxMsgSize

Previous changes
----------------

3.1.1-as10 27-09-04 ! Fix RouteFilebox error. FTrack reported "can't route attached files" if
                      message didn't contain attach.
		    ! Fix CONFIG variable in djgpp and mingw32 makefiles
		    # Now builds on Solaris 9 (SPARC)
                    # Now builds on MacOS X 
		    # bsd makefiles rewritten for bsd make
		    ! Fix byte order on Solaris/SPARC and MacOS X (thanks to Serguei Revtov,
		      2:5021/19.1)
		    # Tested build on NetBSD
		    # FreeBSD port added
		    # Manpage added

3.1.1-as9 09-05-04 ! Fix broken ENABLE_SYSLOG_LOG_FORMAT
		   # Now all documentation placed in one directory. Some filename changes:

		     doc/eng/readme     -> doc/README
		     doc/eng/ftrack.faq -> doc/FAQ
		     doc/eng/report.err -> doc/report.err
		     
		     doc/rus/readme     -> doc/README.ru
		     doc/rus/ftrack.faq -> doc/FAQ.ru
		     doc/rus/report.err -> doc/report.err.ru
		     
		     All russian text now in CP1251
		     
		   # Now you can choose config path by setting FTRACKCONFIG environment
		     variable.
		   # New action: "Action: RouteHub"
		   ! Code cleanup
		   ! Fix missing BSO directory creation
		   ! Fix crash when ftrack detects bad date field in message.
		   # New perl API function: FindHub() which returns hub number for specified node.
		   ! Fix build with Watcom C/C++ compiler (tested with OpenWatcom 1.2).
		   # smapi updated to version 2.4
		   # Now builds on OpenBSD
		   ! Fix ignore attach path by default.
		   # New keyword: IgnoreAttachPath.
		   ! Move attaches to fileboxes (when UseFileBoxes set or used Action: RouteFilebox) 
		     only if KFS flag present. In other case copy attaches.

3.1.1-as8 17-03-04 ! Removed Str class. Returned string processing code from older versions.
                   ! Fix build on FreeBSD 4.x with ENABLE_SCRIPTS=1.
		   ! Fix DOS path stripping under *NIX and vice versa when processing attaches
		   ! Don't exit if can't CopyAttach or MoveAttach
		   # Now messages with attaches placed to filebox as individual packets if UseFileBoxes
                   # New action: "Action: RouteFilebox". Works same as "Action: Route" but
		     message placed as individual packet to filebox.
		   # New actions: "Action: ToLowerPath" and "Action: ToUpperPath" which
		     converts path of attached files and filename to lower or upper case.
		   # New actions: "Action: CopyAttachFilebox" and "Action: MoveAttachFilebox"
		     which copies or moves attach to filebox.

3.1.1-as7 20-02-04 ! Remove unused code.
		   # Added filebox support. New keywords: 

		    UseFileBoxes
		    FileBoxDir: <Dir>
		    FileBoxType: <Type>
		    
		    See documentation for details.
		    
		   # Added Makefile for cygwin (thanks to Dmitry Pankov, 2:5022/81)
		   # Now you can choose actions when FTrack detect bad packet. New keyword:
		   
		     BadPackets: <mode>
		   
		   ! Fix errors when compile with MinGW32
		   ! Fix gcc3 compilation warnings
		     
3.1.1-as6 04-02-04 # New keys for %FMSG hash: InetFromAddr and InetToAddr.
		   ! script.cpp: Fix key size for AreaName
		   # Scripting now full documented.
		   # Now you can change with Update() API function new keys of %FMSG hash:
		     FromAddr, ToAddr, AreaName, Body, Cost, ReplyTo, NextReply and TimesRead.
		   # For compartible returned old log format. It's using by default.
		     To enable syslog-like format run make with ENABLE_SYSLOG_LOG_FORMAT=1 option.
		     To enable PID logging (for both formats) run make with
		     ENABLE_LOG_PID=1 option

3.1.1-as5 02-02-04 # ficl scripts -> perl scripts (backport from FTrack 4.0.1). 
		     Scripts disabled by default. To enable uncomment
		     ENABLE_SCRIPTS=1 in Makefile. WARNING! Code _NOT_ fully tested!
		     Documentation will be available later. See sample script for now.
		   # Default config file now can changed from Makefile (see CONFIG variable)

3.1.1-as4 31-01-04 # New template macros: @InetFromAddr which contain sender
		     address in internet-style (p.f.n.z)
		   # New template macros: @InetToAddr which contain recepient
		     address in internet-style (p.f.n.z)
		   ! Fixed product version in packets
		   # Now smapi code included in project

trunk/src/rntrack.cpp
/*
 *  RNtrack - FTN message tracker/router
 *
 *  rntrack.cpp - Main module
 *
 *  Copyright (c) 2003-2005 Alex Soukhotine, 2:5030/1157
 *	
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  $Id: rntrack.cpp,v 1.2 2005/01/15 02:12:58 ph0enix Exp $
 */

#include <locale.h>
#ifndef UNIX
#include <io.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#if !defined (UNIX) && !defined(__DJGPP__)
#include <sys/utime.h>
#else
#include <sys/types.h>
#include <utime.h>
#include <errno.h>
#include <unistd.h>
#endif
#ifdef __SUN__
#include <time.h>
#endif
#include "constant.hpp"
#include "vars.hpp"
#include "help.hpp"
#include "getopt.hpp"
#include "configure.hpp"
#include "log.hpp"
#include "nodelist.hpp"
#include "msg.hpp"
#include "msgbase.hpp"
#include "scandir.hpp"
#include "aka.hpp"
#include "passwd.hpp"
#include "parsetpl.hpp"
#include "tmstamp.hpp"
#include "script.hpp"
#include <smapi/msgapi.h>

static struct utimbuf ut;
static struct _minf mnf;

// --------------------------------------------------------------------
//                         Signals catch functions
// --------------------------------------------------------------------
void AbrtHndl(int Sign) {
FILE *fh;

   printf("Internal Error!!! Please, read report.err in documentation.\n");
   printf("---------------------------------------\n");
   printf("%s RNtrack %s\n",Date(),ProgVersion);
   printf("Check point: %ld\n",(long int)CHP);
   printf("Signal: %u\n",Sign);
   printf("---------------------------------------\n");
   fh = fopen("RNtrack.err","a+t");
   if (fh != NULL) {
      fprintf(fh,"Internal Error!!! Please, read report.err in documentation.\n");   
      fprintf(fh,"---------------------------------------\n");
      fprintf(fh,"%s RNtrack %s\n",Date(),ProgVersion);
      fprintf(fh,"Check point: %ld\n",(long int)CHP);
      fprintf(fh,"Signal: %u\n",Sign);
      fprintf(fh,"---------------------------------------\n");
   }
   fclose(fh);
   ReleaseSemaphore();
   exit(-1);
}

// --------------------------------------------------------------------
//                            Init all subsystems.
// --------------------------------------------------------------------
int InitSystem(void) {

#ifndef _DEBUG
   CHP = 99100;
   signal(SIGSEGV,&AbrtHndl);
   signal(SIGILL,&AbrtHndl);
#endif

   if (sizeof(int) != 4 || sizeof(long) != 4) {
      fprintf(stderr,"RNtrack compiled for wrong platform.\nPlease, call to author.\n");
      if (sizeof(long) == 8 && sizeof(void *) == 8 && sizeof(int) == 4) {
          fprintf(stderr,"Continue at your own risk! This software is not 64-bit ready yet!\n");
      } else {
          return FALSE;
      }
   }
   CHP = 99101;
   tzset();
   CHP = 99102;
   Log.LogLevel(2);
   memset(&mnf, '\0', sizeof (mnf));
   CHP = 99103;
   mnf.req_version = 1;
   MsgOpenApi(&mnf);
   CHP = 99104;


   if (InitScriptSystem() != TRUE) {
      Log.Level(LOGE) << "Unable to start script system" << EOL;
	  return FALSE;
   }
   if (ConfigFile == NULL) {
      ConfigFile = strdup(DefaultConfig);
   }
   CHP = 99105;

   if (ParseConfig(ConfigFile) != 0) {
      CHP = 9910501;
      return FALSE;
   }

   CHP = 99106;
   if (!MyAddr.Valid()) {
      Log.Level(LOGE) << "Error. Own Address is not defined." << EOL;
      return FALSE;
   }
   if (ForceINTL) {
      UseOwnZone = TRUE;
   }

   if (!Ndl.Load()) {
      return FALSE;
   }

   if (!InitOutbound()) {
      return FALSE;
   }

   if (!SetSemaphore()) {
      return FALSE;
   }

   InitScriptValues();

   switch (DoSomeWordRc("BeforeWork")) {
      case SS_ERROR: return FALSE;
      default: break;
   }

   return TRUE;
}

// --------------------------------------------------------------------
//                         Shutdown all subsystems.
// --------------------------------------------------------------------
void DoneSystem(void) {
int i;

//   ScanDirs.clearAndDestroy();
   DoSomeWord("AfterWork");
   CHP = 99200;
   DestroyAka();
   CHP = 99201;
   DestroyPasswd();
   CHP = 99202;
   StopScriptSystem();
   if (Log.Opened() && LogLevel != 0) {
      CHP = 99203;
      Log.Level(LOGE) << "--- RNtrack " << ProgVersion << " stopped." << EOL;
      Log.Close();
   }
   CHP = 99204;
   if (TimeStampFile != NULL) {
      CHP = 99205;
      if (utime(TimeStampFile,&ut) != 0) {
         CHP = 99206;
         i = errno;
         Log.Level(LOGE) << "Unable to set modification time for file '" << TimeStampFile << "'. Errno: " << i << EOL;
         CHP = 99207;
      }
      CHP = 99208;
   }
   ReleaseSemaphore();
   CHP = 99209;
}

// --------------------------------------------------------------------
//                              M A I N
// --------------------------------------------------------------------
int main (int argc, char *argv[]) {
int Option;
int DoScan;
char *tmpCf;

   setlocale(LC_CTYPE, "");
   CHP = 99000;
   ut.actime = time(NULL);
   ut.modtime = time(NULL);
   Hello();

   DoScan = FALSE;
   CHP = 99001;

   tmpCf = getenv("RNTRACKCONFIG");
   
   if (tmpCf)
    ConfigFile = strdup(tmpCf);
   
   while ((Option = ParseCmdArgs(argc, argv, "c:uh?t")) != EOF) {
      switch (Option) {
         case 'u' : UnpackNeed = TRUE;
                    break;
         case 'c' : /* Kill all mail */
                    ConfigFile = strdup(optionArg);
                    break;
         case 't' : DoScan = TRUE;
                    break;
         case '?' :
         case 'h' : Help(); exit(0);
      } /* switch */
   } /* while */

// ---
   atexit(&DoneSystem);

   CHP = 99002;
   if (!InitSystem()) {
      return (-1);
   }
   CHP = 99003;
   if (DoScan) {
      LastTime = 0;
   }
// ---

   if (UnpackNeed) {
      if (!DoRepack()) {
         return (-1);
      }
   }

   if (LogLevel >= 5) {
      PrintScanDirs();
   }

   DoScanDirs();

   return (0);
}

// ---------------------------- END --------------------------------------
